### PR TITLE
Move freeing of an old enc_write_ctx/write_hash to dtls1_clear_sent_buffer (3.1/3.0)

### LINKED ISSUE #22664: DTLS use-after-free triggered by error injection
If a memory error happens here we encounter a use-after-free:
```
diff --git a/ssl/statem/statem_dtls.c b/ssl/statem/statem_dtls.c
index a88b0dfeac..539f6a3679 100644
--- a/ssl/statem/statem_dtls.c
+++ b/ssl/statem/statem_dtls.c
@@ -1172,7 +1172,8 @@ int dtls1_buffer_message(SSL_CONNECTION *s, int is_ccs)
          char)(dtls1_get_queue_priority(frag->msg_header.seq,
                                         frag->msg_header.is_ccs));
 
-    item = pitem_new(seq64be, frag);
+    //item = pitem_new(seq64be, frag);
+    item = is_ccs ? NULL : pitem_new(seq64be, frag);
     if (item == NULL) {
         dtls1_hm_fragment_free(frag);
         return 0;
```

```
70-test_bad_dtls.t ...................... 
=================================================================
==21913==ERROR: AddressSanitizer: heap-use-after-free on address 0x62100000a790 at pc 0x7f049537bd5f bp 0x7ffe5206f7c0 sp 0x7ffe5206f7b8
READ of size 8 at 0x62100000a790 thread T0
    #0 0x7f049537bd5e in tls_write_records ssl/record/methods/tls_common.c:1864
    #1 0x7f049536592e in do_dtls1_write ssl/record/rec_layer_d1.c:659
    #2 0x7f049526ed39 in dtls1_dispatch_alert ssl/d1_msg.c:57
    #3 0x7f04953a4436 in ossl_statem_fatal ssl/statem/statem.c:170
    #4 0x7f04953a5050 in write_state_machine ssl/statem/statem.c:934
    #5 0x7f04953a5050 in state_machine ssl/statem/statem.c:487
    #6 0x7f04952a03f4 in SSL_do_handshake ssl/ssl_lib.c:4731
    #7 0x55a5971ded16 in test_bad_dtls test/bad_dtls_test.c:555
    #8 0x55a5971e00cc in run_tests test/testutil/driver.c:342
    #9 0x55a5971dccad in main test/testutil/main.c:31
    #10 0x7f04944461c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #11 0x7f0494446284 in __libc_start_main_impl ../csu/libc-start.c:360
    #12 0x55a5971dce60 in _start (/home/ed/ossl/openssl/test/bad_dtls_test+0xbe60)
0x62100000a790 is located 1680 bytes inside of 4432-byte region [0x62100000a100,0x62100000b250)
freed by thread T0 here:
    #0 0x7f04956b76a8 in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x7f049537a7bd in tls_free ssl/record/methods/tls_common.c:1462
    #2 0x7f049536dabe in dtls_free ssl/record/methods/dtls_meth.c:623
    #3 0x7f04953bbf62 in dtls1_hm_fragment_free ssl/statem/statem_dtls.c:103
    #4 0x7f04953bea2f in dtls1_buffer_message ssl/statem/statem_dtls.c:1178
    #5 0x7f04953c2528 in dtls1_close_construct_packet ssl/statem/statem_dtls.c:1360
    #6 0x7f04953a4dc9 in write_state_machine ssl/statem/statem.c:910
    #7 0x7f04953a4dc9 in state_machine ssl/statem/statem.c:487
    #8 0x7f04952a03f4 in SSL_do_handshake ssl/ssl_lib.c:4731
    #9 0x55a5971ded16 in test_bad_dtls test/bad_dtls_test.c:555
    #10 0x55a5971e00cc in run_tests test/testutil/driver.c:342
    #11 0x55a5971dccad in main test/testutil/main.c:31
    #12 0x7f04944461c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
previously allocated by thread T0 here:
    #0 0x7f04956b89cf in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7f0494aea990 in CRYPTO_malloc crypto/mem.c:202
    #2 0x7f0494aeab30 in CRYPTO_zalloc crypto/mem.c:222
    #3 0x7f049537feb3 in tls_int_new_record_layer ssl/record/methods/tls_common.c:1235
    #4 0x7f049536dcff in dtls_new_record_layer ssl/record/methods/dtls_meth.c:643
    #5 0x7f049536b903 in ssl_set_new_record_layer ssl/record/rec_layer_s3.c:1363
    #6 0x7f049529290d in clear_record_layer ssl/ssl_lib.c:560
    #7 0x7f0495293bf5 in ossl_ssl_connection_reset ssl/ssl_lib.c:661
    #8 0x7f04953a5a0e in state_machine ssl/statem/statem.c:377
    #9 0x7f04952a03f4 in SSL_do_handshake ssl/ssl_lib.c:4731
    #10 0x55a5971de7bb in test_bad_dtls test/bad_dtls_test.c:535
    #11 0x55a5971e00cc in run_tests test/testutil/driver.c:342
    #12 0x55a5971dccad in main test/testutil/main.c:31
    #13 0x7f04944461c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
SUMMARY: AddressSanitizer: heap-use-after-free ssl/record/methods/tls_common.c:1864 in tls_write_records
Shadow bytes around the buggy address:
  0x0c427fff94a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff94b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff94c0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff94d0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff94e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c427fff94f0: fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff9500: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff9510: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff9520: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff9530: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c427fff9540: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==21913==ABORTING 
```

A bug-fix is not obvious at all.

### PR DESCRIPTION
**Pull request by mattcaswell**
When we are clearing the sent messages queue we should ensure we free any
old enc_write_ctx/write_hash that are no longer in use. Previously this
logic was in dtls1_hm_fragment_free() - but this can end up freeing the
current enc_write_ctx/write_hash under certain error conditions.
    
Fixes #22664

We also fix an issue during allocation of a new hm_fragment to ensure the data is initialised (commit from @nhorman originall in PR https://github.com/openssl/openssl/pull/22677)

### CODE REVIEW

**Comment by mattcaswell** [comment]
This is equivalent to #22679 but for the 3.1/3.0 branches

**Comment by mattcaswell** [comment]
CI failures are due to a break in 3.1 fixed by #22682

**Review by t8m — APPROVED** [review]

File: ssl/d1_lib.c [hunk 1]
```diff
@@ -130,6 +130,23 @@ void dtls1_clear_sent_buffer(SSL *s)
 
     while ((item = pqueue_pop(s->d1->sent_messages)) != NULL) {
         frag = (hm_fragment *)item->data;
+
+        if (frag->msg_header.is_ccs) {
+            /*
```
Comment (nhorman) [inline ssl/d1_lib.c:135]: nit: comment looks misaligned

**Review by nhorman — APPROVED** [review]
with the misaligned comment fixed

**Review by t-j-h — APPROVED** [review]

**Comment by mattcaswell** [comment]
Just pushed a trivial fixup to address the alignment issue. I will assume all approvals still hold.

**Comment by bernd-edlinger** [comment]
Looks like this does not fix the issue at hand:

e963b652fd5d7deb7ff40d4282f95bd9b7b3f07d with following mod

```
diff --git a/ssl/t1_enc.c b/ssl/t1_enc.c
index 91238e6457..cd30e53956 100644
--- a/ssl/t1_enc.c
+++ b/ssl/t1_enc.c
@@ -293,7 +293,7 @@ int tls1_change_cipher_state(SSL *s, int which)
         }
         dd = s->enc_write_ctx;
         if (SSL_IS_DTLS(s)) {
-            mac_ctx = EVP_MD_CTX_new();
+            mac_ctx = s->write_hash ? NULL : EVP_MD_CTX_new();
             if (mac_ctx == NULL) {
                 SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_MALLOC_FAILURE);
                 goto err;
```

still does this use-after-free:
```            1..9
            # ERROR: (int) 'peer->status == PEER_RETRY' failed @ test/helpers/handshake.c:772
            # [2] compared to [1]
=================================================================
==10359==ERROR: AddressSanitizer: heap-use-after-free on address 0x607000011fa8 at pc 0x7faf3e463416 bp 0x7fff60f256f0 sp 0x7fff60f256e8
READ of size 8 at 0x607000011fa8 thread T0
    #0 0x7faf3e463415 in EVP_MD_CTX_test_flags crypto/evp/evp_lib.c:1096
    #1 0x7faf3e4298c2 in evp_md_ctx_reset_ex crypto/evp/digest.c:85
    #2 0x7faf3e4298c2 in evp_md_ctx_reset_ex crypto/evp/digest.c:75
    #3 0x7faf3e4298c2 in EVP_MD_CTX_reset crypto/evp/digest.c:101
    #4 0x7faf3e429955 in EVP_MD_CTX_free crypto/evp/digest.c:140
    #5 0x7faf3e429955 in EVP_MD_CTX_free crypto/evp/digest.c:135
    #6 0x7faf3eca50c9 in dtls1_clear_sent_buffer ssl/d1_lib.c:146
    #7 0x7faf3eca51c0 in dtls1_clear_queues ssl/d1_lib.c:111
    #8 0x7faf3eca51c0 in dtls1_free ssl/d1_lib.c:163
    #9 0x7faf3ecdf59a in SSL_free ssl/ssl_lib.c:1250
    #10 0x55a9b20f5ebe in peer_free_data test/helpers/handshake.c:761
    #11 0x55a9b20f5ebe in do_handshake_internal test/helpers/handshake.c:1720
    #12 0x55a9b20f775a in do_handshake test/helpers/handshake.c:1732
    #13 0x55a9b20fdada in test_handshake test/ssl_test.c:525
    #14 0x55a9b2105338 in run_tests test/testutil/driver.c:370
    #15 0x55a9b20f028d in main test/testutil/main.c:31
    #16 0x7faf3de461c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #17 0x7faf3de46284 in __libc_start_main_impl ../csu/libc-start.c:360
    #18 0x55a9b20f03d0 in _start (/home/ed/ossl/openssl/test/ssl_test+0x123d0)
0x607000011fa8 is located 24 bytes inside of 72-byte region [0x607000011f90,0x607000011fd8)
freed by thread T0 here:
    #0 0x7faf3eeb76a8 in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x7faf3ecd8f63 in ssl_clear_hash_ctx ssl/ssl_lib.c:4802
    #2 0x7faf3ecdf15b in clear_ciphers ssl/ssl_lib.c:567
    #3 0x7faf3ecdf15b in SSL_free ssl/ssl_lib.c:1215
    #4 0x55a9b20f5ebe in peer_free_data test/helpers/handshake.c:761
    #5 0x55a9b20f5ebe in do_handshake_internal test/helpers/handshake.c:1720
    #6 0x55a9b20f775a in do_handshake test/helpers/handshake.c:1732
    #7 0x55a9b20fdada in test_handshake test/ssl_test.c:525
    #8 0x55a9b2105338 in run_tests test/testutil/driver.c:370
    #9 0x55a9b20f028d in main test/testutil/main.c:31
    #10 0x7faf3de461c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
previously allocated by thread T0 here:
    #0 0x7faf3eeb89cf in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7faf3e4bf890 in CRYPTO_zalloc crypto/mem.c:203
    #2 0x7faf3ecf2674 in tls1_change_cipher_state ssl/t1_enc.c:296
    #3 0x7faf3ed5c0ea in ossl_statem_client_post_work ssl/statem/statem_clnt.c:810
    #4 0x7faf3ed4b7cc in write_state_machine ssl/statem/statem.c:893
    #5 0x7faf3ed4b7cc in state_machine ssl/statem/statem.c:451
    #6 0x7faf3ecd693f in SSL_do_handshake ssl/ssl_lib.c:3938
    #7 0x55a9b20f2f10 in do_handshake_step test/helpers/handshake.c:775
    #8 0x55a9b20f2f10 in do_handshake_step test/helpers/handshake.c:770
    #9 0x55a9b20f2f10 in do_connect_step test/helpers/handshake.c:1128
    #10 0x55a9b20f68fb in do_handshake_internal test/helpers/handshake.c:1538
    #11 0x55a9b20f775a in do_handshake test/helpers/handshake.c:1732
    #12 0x55a9b20fdada in test_handshake test/ssl_test.c:525
    #13 0x55a9b2105338 in run_tests test/testutil/driver.c:370
    #14 0x55a9b20f028d in main test/testutil/main.c:31
    #15 0x7faf3de461c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
SUMMARY: AddressSanitizer: heap-use-after-free crypto/evp/evp_lib.c:1096 in EVP_MD_CTX_test_flags
Shadow bytes around the buggy address:
  0x0c0e7fffa3a0: fd fd fd fd fd fd fd fa fa fa fa fa fd fd fd fd
  0x0c0e7fffa3b0: fd fd fd fd fd fa fa fa fa fa fd fd fd fd fd fd
  0x0c0e7fffa3c0: fd fd fd fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c0e7fffa3d0: fd fa fa fa fa fa fd fd fd fd fd fd fd fd fd fa
  0x0c0e7fffa3e0: fa fa fa fa fd fd fd fd fd fd fd fd fd fa fa fa
=>0x0c0e7fffa3f0: fa fa fd fd fd[fd]fd fd fd fd fd fa fa fa fa fa
  0x0c0e7fffa400: fd fd fd fd fd fd fd fd fd fa fa fa fa fa fd fd
  0x0c0e7fffa410: fd fd fd fd fd fd fd fa fa fa fa fa fd fd fd fd
  0x0c0e7fffa420: fd fd fd fd fd fa fa fa fa fa fd fd fd fd fd fd
  0x0c0e7fffa430: fd fd fd fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c0e7fffa440: fd fa fa fa fa fa fd fd fd fd fd fd fd fd fd fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==10359==ABORTING
../../util/wrap.pl ../../test/ssl_test 18-dtls-renegotiate.cnf.default default => 1
    not ok 6 - running CTLOG_FILE=test/ct/log_list.cnf TEST_CERTS_DIR=test/certs test/ssl_test test/ssl-tests/18-dtls-renegotiate.cnf default
    #   Failed test 'running CTLOG_FILE=test/ct/log_list.cnf TEST_CERTS_DIR=test/certs test/ssl_test test/ssl-tests/18-dtls-renegotiate.cnf default'
    #   at test/recipes/80-test_ssl_new.t line 185.
    # Looks like you failed 2 tests of 6.
not ok 18 - Test configuration 18-dtls-renegotiate.cnf

```

**Comment by mattcaswell** [comment]
Fixup pushed that should hopefully address this.

@t8m, @t-j-h: please reconfirm

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Still needs a second approval or reconfirmation from @t-j-h.

**Comment by mattcaswell** [comment]
Ping for second approval @openssl/committers!

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by mattcaswell** [comment]
Thanks @beldmit - could you also look at the very similar #22679?

**Review by t-j-h — APPROVED** [review]
Sorry - missed the re-approve ping ...

**Comment by mattcaswell** [comment]
Pushed to 3.1 and 3.0

**Pull request closed without merging**